Analysis/src/GenParticle.cpp

```diff
@@ -0,0 +1,235 @@
+
+    if (code_iter != particleCodeMap.end()){
+        for (const GenParticle* particle : code_iter->second){
+            if(requireIsLastCopy && !particle->genStatusFlags.isLastCopy()) continue;
+            results.insert(particle);
+        }
+    }
+    return results;
+}
+
+void GenEvent::GetTypesParticles(const std::set<particles::ParticleCode::ParticleType>& type_names, const GenParticle* mother, GenParticleSet& result) const
+{
+    if(mother->genStatusFlags.isPrompt() && mother->genStatusFlags.isLastCopy() && particle_types.count(mother->pdg)
+        && type_names.count(particle_types.at(mother->pdg))) {
+//        if(shared && areParented(mother,mother->daughters.at(0)) && areParented(mother,mother->daughters.at(1)) )
+        result.insert(mother);
+    } else {
+        for(const GenParticle* daughter : mother->daughters)
+            GetTypesParticles(type_names, daughter, result);
+    }
+}
+
+bool GenEvent::areParented(const GenParticle* daughter, const GenParticle* mother) const
+{
+    for (size_t i = 0; i < daughter->mothers.size(); i++) {
+        const GenParticle* mother_ = daughter->mothers.at(i);
+        if(mother_ == mother)
+            return true;
+        if(areParented(mother_,mother))
+            return true;
+    }
+    return false;
+}
+
+std::string& GenEvent::GetParticleName(int pdgId) const
+{
+    auto iter = particle_names.find(pdgId);
+    if(iter == particle_names.end()) throw exception("Name not found for particle with pdgId = %1%") % pdgId;
+    return iter->second;
+}
+
+void GenEvent::intializeNames(const std::string& fileName)
+{
+    particle_names.clear();
+    particle_types.clear();
+    particle_charge.clear();
+    std::ifstream f(fileName.c_str());
+    if(!f.is_open() )
+        throw analysis::exception("Unable to read the configuration file ");
+
+    while(f.good()) {
+        std::string line;
+        std::getline(f, line);
+        if(line.empty() || line[0] == '#')
+            continue;
+        auto pdg_name_type = SplitValueList(line, false, ",");
+        if(pdg_name_type.size() < 2)
+            throw exception("Invalid particle definition (1): '%1%'") % line;
+
+        const int pdgId = Parse<int>(pdg_name_type.at(0));
+        const std::string& name = pdg_name_type.at(1);
+        if(particle_names.count(pdgId))
+            throw exception("Duplicated definition of particle with pdgId = %1%") % pdgId;
+        particle_names[pdgId] = name;
+
+        if(pdg_name_type.size() > 2){
+
+            const std::string& type_str = pdg_name_type.at(2);
+
+            std::map<std::string, particles::ParticleCode::ParticleType> ParticleTypeMap;
+            if(type_str == "baryon")
+                ParticleTypeMap[type_str] = particles::ParticleCode::ParticleType::baryon;
+            else if(type_str == "meson")
+                ParticleTypeMap[type_str] = particles::ParticleCode::ParticleType::meson;
+
+            if(ParticleTypeMap.count(type_str)){
```

Comment:
Don't create local `ParticleTypeMap`. Use Parse.

Reply:
I'm not sure how to use Parse in here. I try something like it's commented below

Reply:
why Parse didn't work?

Reply:
Doing it like I put on the commented code below it gives this errors:

> error: invalid operands to binary expression ('std::__1::basic_istream<char>' and 'particles::ParticleType')

> note: in instantiation of member function 'analysis::detail::StringIOImpl<particles::ParticleType, char>::TryParse'
>       requested here
>     return detail::StringIOImpl<typename std::remove_reference<T>::type, CharT>::TryParse(str, t);

Reply:
Try to add in Particle.h (inside particles namespace) following lines:
```
using ::analysis::operator<<;
using ::analysis::operator>>;
```

Reply:
it gives this error now: 

> ERROR: Parse of string 'particles::ParticleType::baryon' to N9particles12ParticleTypeE is failed.

Reply:
So, it compiles, but gives you an error when running?
show me the full code

Reply:
Yes, after compile throws that. The code  bellow 
           
```
const std::string& type_str = pdg_name_type_charge.at(2);
std::ostringstream ss_type;
 ss_type << "particles::ParticleType::"<< type_str;
std::string type = ss_type.str();
particle_types[pdgId] = Parse<particles::ParticleType>(type);


```

Reply:
the code should be:
```
particle_types[pdgId] = Parse<particles::ParticleType>(pdg_name_type_charge.at(2));
```